src/combinatorics/additive/behrend.lean

```diff
@@ -0,0 +1,192 @@
+by simp [map, fin.sum_univ_succ, pow_succ', ←mul_assoc, ←sum_mul]
+
+lemma map_succ' (a : fin (n + 1) → ℕ) : map d a = a 0 + map d (a ∘ fin.succ) * d := map_succ _
+
+lemma map_monotone (d : ℕ) : monotone (map d : (fin n → ℕ) → ℕ) :=
+λ x y h, by { dsimp, exact sum_le_sum (λ i _, nat.mul_le_mul_right _ $ h i) }
+
+lemma map_mod (a : fin n.succ → ℕ) : map d a % d = a 0 % d :=
+by rw [map_succ, nat.add_mul_mod_self_right]
+
+lemma map_eq_iff {x₁ x₂ : fin n.succ → ℕ} (hx₁ : ∀ i, x₁ i < d) (hx₂ : ∀ i, x₂ i < d) :
+  map d x₁ = map d x₂ ↔ x₁ 0 = x₂ 0 ∧ map d (x₁ ∘ fin.succ) = map d (x₂ ∘ fin.succ) :=
+begin
+  refine ⟨λ h, _, λ h, by rw [map_succ', map_succ', h.1, h.2]⟩,
+  have : x₁ 0 = x₂ 0,
+  { rw [←mod_eq_of_lt (hx₁ _), ←map_mod, ←mod_eq_of_lt (hx₂ _), ←map_mod, h] },
+  rw [map_succ, map_succ, this, add_right_inj, mul_eq_mul_right_iff] at h,
+  exact ⟨this, h.resolve_right (pos_of_gt (hx₁ 0)).ne'⟩,
+end
+
+lemma map_inj_on : {x : fin n → ℕ | ∀ i, x i < d}.inj_on (map d) :=
+begin
+  intros x₁ hx₁ x₂ hx₂ h,
+  induction n with n ih,
+  { simp },
+  ext i,
+  have x := (map_eq_iff hx₁ hx₂).1 h,
+  refine fin.cases x.1 (congr_fun $ ih (λ _, _) (λ _, _) x.2) i,
+  { exact hx₁ _ },
+  { exact hx₂ _ }
+end
+
+lemma map_le_of_mem_box (hx : x ∈ box n d) :
+  map (2 * d - 1) x ≤ ∑ i : fin n, (d - 1) * (2 * d - 1) ^ (i : ℕ) :=
+map_monotone (2 * d - 1) $ λ _, nat.le_pred_of_lt $ mem_box.1 hx _
+
+lemma add_salem_spencer_sphere : add_salem_spencer (sphere n d k : set (fin n → ℕ)) :=
+begin
+  set f : (fin n → ℕ) →+ euclidean_space ℝ (fin n) :=
+  { to_fun := λ f, (coe : ℕ → ℝ) ∘ f,
+    map_zero' := rfl,
+    map_add' := λ _ _, funext $ λ _, cast_add _ _ },
+  refine add_salem_spencer.of_image (f.to_add_freiman_hom (sphere n d k) 2) _ _,
+  { exact cast_injective.comp_left.inj_on _ },
+  refine (add_salem_spencer_sphere 0 $ sqrt k).mono (set.image_subset_iff.2 $ λ x, _),
+  rw [set.mem_preimage, mem_sphere_zero_iff_norm],
+  exact norm_of_mem_sphere,
+end
+
+lemma add_salem_spencer_image_sphere :
+  add_salem_spencer ((sphere n d k).image (map (2 * d - 1)) : set ℕ) :=
+begin
+  rw coe_image,
+  refine @add_salem_spencer.image _ (fin n → ℕ) ℕ _ _ (sphere n d k) _ (map (2 * d - 1))
+   (map_inj_on.mono _) add_salem_spencer_sphere,
+  rw set.add_subset_iff,
+  rintro a ha b hb i,
+  have hai := mem_box.1 (sphere_subset_box ha) i,
+  have hbi := mem_box.1 (sphere_subset_box hb) i,
+  rw [lt_tsub_iff_right, ←succ_le_iff, two_mul],
+  exact (add_add_add_comm _ _ 1 1).trans_le (add_le_add hai hbi),
+end
+
+lemma sum_sq_le_of_mem_box (hx : x ∈ box n d) : ∑ i : fin n, (x i)^2 ≤ n * (d - 1)^2 :=
+begin
+  rw mem_box at hx,
+  have : ∀ i, x i ^ 2 ≤ (d - 1) ^ 2 := λ i, nat.pow_le_pow_of_le_left (nat.le_pred_of_lt (hx i)) _,
+  exact (sum_le_card_nsmul univ _ _ $ λ i _, this i).trans (by rw [card_fin, smul_eq_mul]),
+end
+
+lemma digits_sum_eq : ∑ i : fin n, (d - 1) * (2 * d - 1) ^ (i : ℕ) = ((2 * d - 1) ^ n - 1) / 2 :=
```

Comment:
It's not obvious to me where the word "digits" comes from in this lemma and the one below - can you add a docstring?